File: skypydb-js/src/database/mixins/vector/update_items.ts
import { CollectionNotFoundError } from "../../../errors";
import { InputValidator } from "../../../security/validation";
import type { EmbeddingMatrix, Metadata } from "../../../types";
import { AddItemsMixin } from "./add_items";

export abstract class UpdateItemsMixin extends AddItemsMixin {
  async update(
    collection_name: string,
    ids: string[],
    embeddings?: EmbeddingMatrix,
    documents?: string[],
    metadatas?: Metadata[]
  ): Promise<void> {
    const validated = InputValidator.validate_table_name(collection_name);
    if (!this.collection_exists(validated)) {
      throw new CollectionNotFoundError(`Collection '${validated}' not found`);
    }

    let resolved_embeddings = embeddings;
    if (!resolved_embeddings && documents) {
      if (!this.embedding_function) {
        throw new Error("Documents provided but no embedding function set.");
      }
      resolved_embeddings = await this.embedding_function(documents);
    }

    for (let index = 0; index < ids.length; index += 1) {
      const updates: string[] = [];
      const params: unknown[] = [];

      if (resolved_embeddings) {
        updates.push("embedding = ?");
        params.push(JSON.stringify(resolved_embeddings[index]));
      }
      if (documents) {
        updates.push("document = ?");
        params.push(documents[index]);
      }
      if (metadatas) {
        updates.push("metadata = ?");
        params.push(metadatas[index] ? JSON.stringify(metadatas[index]) : null);
      }

      if (updates.length > 0) {
        const query = `UPDATE [vec_${validated}] SET ${updates.join(", ")} WHERE id = ?`;
        this.conn.prepare(query).run(...params, ids[index]);
      }
    }
  }
}

File: skypydb/database/mixins/vector/update_items.py
"""Update operations for vector collection records."""

import json
from typing import Any, Dict, List, Optional

from skypydb.security.validation import InputValidator


class UpdateItemsMixin:
    """Provide record updates for vector collections."""

    def update(
        self,
        collection_name: str,
        ids: List[str],
        embeddings: Optional[List[List[float]]] = None,
        documents: Optional[List[str]] = None,
        metadatas: Optional[List[Dict[str, Any]]] = None,
    ) -> None:
        """Update records by ID in a collection."""

        collection_name = InputValidator.validate_table_name(collection_name)
        if not self.collection_exists(collection_name):
            raise ValueError(f"Collection '{collection_name}' not found")

        if embeddings is None and documents is not None:
            if self.embedding_function is None:
                raise ValueError("Documents provided but no embedding function set.")
            embeddings = self.embedding_function(documents)

        cursor = self.conn.cursor()

        for idx, item_id in enumerate(ids):
            updates = []
            params = []

            if embeddings is not None:
                updates.append("embedding = ?")
                params.append(json.dumps(embeddings[idx]))
            if documents is not None:
                updates.append("document = ?")
                params.append(documents[idx])
            if metadatas is not None:
                updates.append("metadata = ?")
                params.append(json.dumps(metadatas[idx]) if metadatas[idx] else None)

            if updates:
                params.append(item_id)
                cursor.execute(
                    f"UPDATE [vec_{collection_name}] SET {', '.join(updates)} WHERE id = ?",
                    params,
                )
        self.conn.commit()
